services/s3/src/it/java/software/amazon/awssdk/services/s3/checksum/HttpChecksumIntegrationTest.java

```diff
@@ -410,4 +409,90 @@ public void syncValidUnsignedTrailerChecksumCalculatedBySdkClient_withHugeFileRe
         assertThat(text).isEqualTo(new String(bytes));
     }
 
+    private static Stream<Arguments> getObjectChecksumValidationParams() {
+        return Stream.of(Arguments.of(true, ChecksumAlgorithm.CRC32, ChecksumMode.ENABLED),
+                         Arguments.of(true, null, ChecksumMode.ENABLED),
+                         Arguments.of(true, ChecksumAlgorithm.CRC32, null),
+                         Arguments.of(true, null, null),
+                         Arguments.of(false, ChecksumAlgorithm.CRC32, ChecksumMode.ENABLED),
+                         Arguments.of(false, null, ChecksumMode.ENABLED),
+                         Arguments.of(false, ChecksumAlgorithm.CRC32, null),
+                         Arguments.of(false, null, null));
+    }
+
+    @ParameterizedTest
+    @MethodSource("getObjectChecksumValidationParams")
+    public void testGetObjectChecksumValidation(boolean checksumValidationEnabled, ChecksumAlgorithm checksumAlgorithm,
+                                                ChecksumMode checksumMode) {
+        S3Client s3 = s3ClientBuilder().overrideConfiguration(o -> o.addExecutionInterceptor(interceptor))
+                                       .serviceConfiguration(S3Configuration.builder()
+                                                                            .checksumValidationEnabled(checksumValidationEnabled)
+                                                                            .build())
+                                       .endpointOverride(URI.create("http://s3." + DEFAULT_REGION + ".amazonaws.com"))
+                                       .build();
+
+        s3.putObject(PutObjectRequest.builder()
+                                     .bucket(BUCKET)
+                                     .key(KEY)
+                                     .checksumAlgorithm(checksumAlgorithm)
+                                     .build(), RequestBody.fromString("Hello world"));
+
+        s3.getObject(GetObjectRequest.builder()
+                                     .bucket(BUCKET)
+                                     .key(KEY)
+                                     .checksumMode(checksumMode)
+                                     .build());
+
+        validateChecksumValidation(checksumValidationEnabled, checksumAlgorithm, checksumMode);
+        interceptor.reset();
+    }
+
+    private void validateChecksumValidation(boolean checksumValidationEnabled, ChecksumAlgorithm checksumAlgorithm,
+                                               ChecksumMode checksumMode) {
+        assertThat(interceptor.responseBody()).isPresent();
+
+        if (checksumValidationEnabled) {
+            if (checksumMode == ChecksumMode.ENABLED) {
```

Comment:
Instead of checking the underlying inputstream, can we check response header, i.e., if  checksum mode is enabled, there shouldn't be any md5 trailer header? This way, we can remove the getter method to retrieve the stream. 

Let's add tests for async as well.